snuba/subscriptions/scheduler.py

```diff
@@ -39,3 +42,42 @@ def find(self, interval: Interval[datetime]) -> Iterator[ScheduledTask[TTask]]:
         ascending order.
         """
         raise NotImplementedError
+
+
+class SubscriptionScheduler(Scheduler[Subscription]):
+    def __init__(
+        self,
+        store: SubscriptionDataStore,
+        partition_id: PartitionId,
+        cache_ttl: timedelta,
+    ) -> None:
+        self.__store = store
+        self.__cache_ttl = cache_ttl
+        self.__partition_id = partition_id
+        self.__subscriptions: List[Subscription] = []
+        self.__last_refresh: Optional[datetime] = None
+
+    def __get_subscriptions(self, current_time: datetime) -> List[Subscription]:
+        if (
+            self.__last_refresh is None
+            or (current_time - self.__last_refresh) > self.__cache_ttl
+        ):
+            self.__subscriptions = [
+                Subscription(SubscriptionIdentifier(self.__partition_id, uuid), data)
+                for uuid, data in self.__store.all()
+            ]
+            self.__last_refresh = current_time
+        return self.__subscriptions
+
+    def find(
+        self, interval: Interval[datetime]
+    ) -> Iterator[ScheduledTask[Subscription]]:
+        subscriptions = self.__get_subscriptions(interval.lower)
+        for i in range(
+            math.ceil(interval.lower.timestamp()),
+            math.ceil(interval.upper.timestamp()),
+        ):
+            for subscription in subscriptions:
+                resolution = int(subscription.data.resolution.total_seconds())
+                if i % resolution == 0:
+                    yield ScheduledTask(datetime.fromtimestamp(i), subscription)
```

Comment:
I wish that we would have called the `resolution` concept something like `frequency` or `period` instead. Resolution sounds too much like the time window to me.

I think this approach is fine and will probably be reasonably easy to optimize later, since we will probably need to time shift these values to avoid bursts on common time alignments. At that point we can store something like:

```python
self.__subscriptions = {
  60: {
    0: [],  # all items to be executed when timestamp % resolution == 0
    …,
    59: [],
  },
  300: {…},
}
```

However we determine that the remainder value should result in items that can be efficiently coalesced together falling into the same slot. I don't think we should worry about that quite yet, but it will probably happen pretty soon. I guess also be aware of this for the UI side — schedules will not always cleanly align on convenient time intervals.

I also think that if we are going to go this route (discretizing time to integers), we should make sure that subscriptions do not have a fractional component in `resolution` when they are instantiated.

Reply:
I don't mind if we want to rename `resolution` on the snuba side, it's pretty low effort.